pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java

```diff
@@ -384,7 +386,12 @@ private void sendOutputMessage(Record srcRecord, Object output) throws Exception
         try {
             this.sink.write(sinkRecord);
         } catch (Exception e) {
-            log.info("Encountered exception in sink write: ", e);
+            if (e instanceof ClassCastException && functionClassLoader != componentClassLoader) {
```

Comment:
Can we do this check at sink create or start?

Reply:
you can't because each record can be different

in Java even if you declare a class with some Generic types, you can always force the system to return something wrong.
so this check must stay here.

Reply:
There is also a question about the type of input for user-defined functions.  In the current implementation, the consumer in the sink sets the schema according to the generic definition of the sink. usually, we use `GenericObject`, If the user-defined function does not match the generic, the current logic will go to the following logic instead of being sent to the sink, is this the expected behavior?


https://github.com/apache/pulsar/blob/5df15dd2edd7eeab309fea35828915c8698ea339/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java#L354-L360

Maybe we need to validate the input parameter types of the user-defined function at the sink start?

Reply:
If you have a Function with the Sink we use the input type of the Function to configure the PulsarSource instead of the Sink type. See [here](https://github.com/apache/pulsar/pull/16740/files#diff-f35e3b19e9f35ef9e868df2020221324c959deb664bff63437bb3a2ddf929397R484-R489).
Does it answer your concerns ?

Reply:
Oh, sorry. I miss this.

However, if the user wants to use it, the input parameter type may need to be carefully defined. In general, users can only use `GenericObject`, which triggers the registration of `auto_consumer` type schema, because only then can multiple topic inputs be processed.

Am I right?

Reply:
You can use the same input type as if you had a regular Sink or Function. If you want to consume any type of Schema then yes, you should use `GenericObject` or `byte[]`. If you know that the Schema will always be String, you can use String as input.